### PR TITLE
test: enable coverage reporting

### LINKED ISSUE #3722: Code coverage
In our TypeScript code, we should add code coverage reporting.
It should be fairly straightforward to set up, as we're already using jest everywhere.

### PR DESCRIPTION
**Pull request by jasonkuhrt**
relates to #3722

### CODE REVIEW

File: src/packages/cli/package.json [hunk 1]
```diff
@@ -127,7 +127,7 @@
     "test:postgresql": "jest integration.postgresql --forceExit --maxConcurrency=10",
     "test:mysql": "jest integration.mysql --forceExit --maxConcurrency=5",
     "test:mariadb": "jest integration.mariadb --forceExit --maxConcurrency=5",
-    "test:commands": "./fixtures/test.sh && jest dotenv- format doctor generate studio && jest introspect --forceExit",
+    "test:commands": "./fixtures/test.sh && jest dotenv- format doctor generate studio introspect --forceExit",
```
[the remark this answers is not in the document]
Comment (Jolg42) [reply src/packages/cli/package.json:130]: Here it was separate so we only forceExit for introspect but it's ok 👍

**Review by Jolg42 — APPROVED** [review]
Awesome!

Do you know what is the current score?

**Comment by jasonkuhrt** [comment]
So a few notes:

- Still working on getting tests to pass... 🤷‍♂️

- I am trying to get this to integrate with codecov. Instead of committing the coverage files to the repo we will rely on keeping that state in codecov. So codecov will be responsible for example for enforcing any diff thresholds of coverage we want.

- Setting up codecov is non-obvious and I'm not sure its working yet. Once we merge a first pass here to trunk I think we'll be working better with it. More PRs might be needed to work out the kinks.

- Once codecov is setup correctly it should start to make comments on our PRs here

- Setting up codecov with a monorepos a bit more complex. It is supposed to be supported, though.

- I am keeping some notes about our coverage setup so [far in notion](https://www.notion.so/prismaio/codecov-ca716b245fcb4a0da5df39b70d2de31c)

- about this:

  > Do you know what is the current score?

  I believe 60-80% for client and cli but I'm not really confident about the scores yet until we finalize the config of everything.

**Comment by jasonkuhrt** [comment]
Ah timeout errors: https://buildkite.com/prisma/prisma2-test/builds/3119#5d211843-6b36-4c61-9cc1-2c4b1456fb89/151-653

Adding coverage reporting slows down tests, unfortunately.

**Comment by Jolg42** [comment]
You got a 
```
The engine commit 29ef6782028310083196853b728a878b78bf1a31 is not yet done. 138 urls are missing. Trying again in 10 seconds
```
SO I think you can try a rerun

We will get rid of that soon probably with external engines package dependency.

**Comment by jasonkuhrt** [comment]
@Jolg42 where do you see that exactly?

**Comment by Jolg42** [comment]
From the Builkite job that failed https://buildkite.com/prisma/prisma2-test/builds/3119#5d211843-6b36-4c61-9cc1-2c4b1456fb89/151-647

**Comment by jasonkuhrt** [comment]
Happy to see the Codecov PR comment kick in :)

Some bad news is that coverage makes tests go slower.

For example sqlite integration tests before [took 1:34m](https://github.com/prisma/prisma/runs/1220435324#step:8:1). Now it [takes 7m](https://github.com/prisma/prisma/pull/3848/checks?check_run_id=1220789780#step:8:593)!

While this is horrible I think this is the way forward... Some things we can look forward to in the future:

- Coverage reporting powered by V8 which includes better perf ([announce](https://jestjs.io/blog/2020/01/21/jest-25#v8-code-coverage), [flag](https://jestjs.io/docs/en/cli#--coverageproviderprovider))
- Coverage reporting based on unit tests only (not integration tests)
- Not limiting concurrency to 1 of course!
- Using more powerful custom machines via a cloud provider

**Comment by jasonkuhrt** [comment]
https://github.com/facebook/jest/issues/9457

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -36,7 +36,7 @@ jobs:
   # CLI-INTEGRATION
   #
   cli-integration:
-    timeout-minutes: 16
+    timeout-minutes: 60
```
Comment (Jolg42) [inline .github/workflows/test.yml:39]: 🙈 Ouch, I hope only temporary? Because that's quite long

**Comment by Jolg42** [comment]
I suggest we remove some files from the coverage analysis because they are not relevant.

We can add this comment on top of these files:
```
/* istanbul ignore file */
```
or use 
```
{
  "jest": {
    "collectCoverageFrom": [
      "src/**/{!(ignore-me),}.ts"
    ]
  }
}
```

Example of not relevant files for me, and that could make the report faster too:
Client
- src/packages/client/src/utils/compileFile.ts
- src/packages/client/src/utils/generateInFolder.ts 
- src/packages/client/src/utils/getTestClient.ts 
- src/packages/client/src/utils/setupMysql.ts 
- src/packages/client/src/utils/setupPostgres.ts 
- src/packages/client/src/runtime/utils/uniqueBy.ts 
- src/packages/client/src/runtime/utils/stringifyObject.ts
- src/packages/client/src/runtime/utils/omit.ts 
- src/packages/client/src/runtime/utils/flatMap.ts 
- src/packages/client/src/runtime/utils/filterObject.ts 
- src/packages/client/src/runtime/utils/deep-set.ts 
- src/packages/client/src/runtime/utils/deep-extend.ts 
- src/packages/client/src/runtime/mergeBy.ts 
- src/packages/client/src/runtime/highlight/** 
- src/packages/client/src/fixtures/**
- src/packages/client/helpers/jestSnapshotSerializer.js

**Comment by jasonkuhrt** [comment]
I will disable coverage reporting for `cli` right now. It is making node run out of memory https://github.com/prisma/prisma/pull/3848/checks?check_run_id=1221460255#step:8:595 it seems. And we know those tests are going to migrate soon anyways. So I'll focus effort here on client package.

File: src/packages/cli/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jasonkuhrt) [reply src/packages/cli/package.json]: Obsolete now that we're ignoring cli in this PR.

File: src/packages/sdk/src/__tests__/getGenerators/getGenerators.test.ts [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import path from 'path'
 import { getGenerators } from '../../getGenerators'
-import { pick } from '../../pick'
 import { omit } from '../../omit'
+import { pick } from '../../pick'
 
-jest.setTimeout(10000)
+jest.setTimeout(20000)
```
Comment (jasonkuhrt) [inline src/packages/sdk/src/__tests__/getGenerators/getGenerators.test.ts:6]: It seems that 10 seconds was not long enough. Example: https://github.com/prisma/prisma/runs/1226432839#step:6:600. So trying 20 seconds...

**Comment by jasonkuhrt** [comment]
All packages except cli should be generating coverage reports now. I left out CLI because of the troubles it was causing with coverage and b/c the tests will undergo major refactoring soon (https://github.com/prisma/prisma/issues/3864).

File: src/packages/cli/package.json [hunk 4]
```diff
@@ -59,7 +59,7 @@
     "prisma2": "build/index.js"
   },
   "prisma": {
-    "version": "f4f4c265b355e72ab29dbb1d4495d4bcaf0e829f",
+    "version": "c4a65524ad3e8b18e8ea0c41a718819b4cd42ee4",
```
Comment (jasonkuhrt) [inline src/packages/cli/package.json:62]: Is it ok for these changes to be committed?

File: src/pnpm-lock.yaml [hunk 5]
```diff
@@ -119,20 +119,20 @@ importers:
...
       '@types/sqlite3': 3.1.6
-      '@types/string-hash': ^1.1.1
-      '@types/verror': ^1.10.4
+      '@types/string-hash': 1.1.1
+      '@types/verror': 1.10.4
       '@types/ws': 7.2.6
       '@typescript-eslint/eslint-plugin': 4.2.0
       '@typescript-eslint/parser': 4.2.0
-      checkpoint-client: 1.1.12
+      checkpoint-client: 1.1.13
```
Comment (jasonkuhrt) [inline src/pnpm-lock.yaml:133]: Hmm... I didn't explicitly do this...

File: src/packages/engine-core/jest.config.js [hunk 6]
```diff
@@ -1,14 +1,10 @@
...
-    'build/',
-    'dist/',
-    'generator/',
-    'runtime/',
-    'scripts/',
-    'sandbox/',
-  ],
+  collectCoverage: true,
+  coverageDirectory: 'src/__tests__/coverage',
+  collectCoverageFrom: ['src/**/*.ts', '!**/__tests__/**/*', '!**/*.test.ts'],
```
Comment (jasonkuhrt) [inline src/packages/engine-core/jest.config.js:6]: The explanation for this:

1. All source files
2. Except for anything inside tests dir (which sometimes contains non-test modules like helper modules)
3. Except all test files (some packages have `__tests__` in multiple places and it seemed this was needed to exclude them all. Thinking about it now it seems like this shouldn't be needed given rule 2 above but... meh.

File: src/pnpm-lock.yaml (hunk 5 above)
Comment (Jolg42) [reply src/pnpm-lock.yaml:133]: We usually don't commit the pnpm-lock in branches, you can ignore it

**Review by timsuchanek — APPROVED** [review]
LGTM

**Pull request merged**